File: _django/loaddata.py
import argparse
import collections
import datetime
import json

from . import bootstrap  # NoQA
from . import models


if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        description='Load a specific fixture, old data will be purged.')
    parser.add_argument('filename', type=str,
                        help='The JSON dataset file')

    args = parser.parse_args()

    User = models.User
    Directors = models.Directors
    Cast = models.Cast
    Person = models.Person
    Movie = models.Movie
    Review = models.Review

    with open(args.filename, 'rt') as f:
        records = json.load(f)

    data = collections.defaultdict(list)
    for rec in records:
        rtype = rec['model'].split('.')[-1]
        datum = rec['fields']
        if 'pk' in rec:
            datum['id'] = rec['pk']
        # convert datetime
        if rtype == 'review':
            datum['creation_time'] = datetime.datetime.fromisoformat(
                datum['creation_time'])

        data[rtype].append(datum)

    batch_size = 1000

    # bulk create all the users
    print(f"Creating {len(data['user'])} users...")
    User.objects.bulk_create(
        [User(**datum) for datum in data['user']],
        batch_size=batch_size
    )

    # bulk create all the people
    print(f"Creating {len(data['person'])} people...")
    Person.objects.bulk_create(
        [Person(**datum) for datum in data['person']],
        batch_size=batch_size
    )

    # bulk create all the movies
    print(f"Creating {len(data['movie'])} movies...")
    Movie.objects.bulk_create(
        [Movie(**datum) for datum in data['movie']],
        batch_size=batch_size
    )

    # bulk create all the reviews
    print(f"Creating {len(data['review'])} reviews...")
    Review.objects.bulk_create(
        [Review(**datum) for datum in data['review']],
        batch_size=batch_size
    )

    # bulk create all the directors
    print(f"Creating {len(data['directors'])} directors...")
    Directors.objects.bulk_create(
        [Directors(**datum) for datum in data['directors']],
        batch_size=batch_size
    )

    # bulk create all the cast
    print(f"Creating {len(data['cast'])} cast...")
    Cast.objects.bulk_create(
        [Cast(**datum) for datum in data['cast']],
        batch_size=batch_size
    )

File: _loopback/server/loaddata.js
// Copyright IBM Corp. 2016. All Rights Reserved.
// Node module: loopback-workspace
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

'use strict';

var loopback = require('loopback');
var boot = require('loopback-boot');
const fs = require('fs');
const _ = require('lodash');
const ProgressBar = require('progress');

var app = module.exports = loopback();

const BATCH_SIZE = 1000;


async function bulk_insert(app, data, modelName) {
  let dataName = modelName.toLowerCase();
  let adj = true;
  var bar = new ProgressBar(
    _.padEnd(modelName, 31, ' ') + '|:bar| :current/:total\t:elapsed sec',
    {total: data[dataName].length, width: 32}
  );
  bar.tick();

  await app.dataSources.psqlDs.automigrate(modelName);

  for (var chunk of _.chunk(data[dataName], BATCH_SIZE)) {
    await app.dataSources.psqlDs.transaction(async models => {
      models[modelName].create(chunk, function(err) {
        if (err) throw err;
      });
    });

    if (adj) {
      bar.tick(chunk.length - 1);
      adj = false;
    } else {
      bar.tick(chunk.length);
    }
  }
}


// Bootstrap the application, configure models, datasources and middleware.
// Sub-apps like REST API are mounted via boot scripts.
boot(app, __dirname, async function(err) {
  if (err) throw err;

  if (process.argv.length != 3) {
    throw new Error(
      'Correct usage: node _loopback/server/loaddata.js dataset.json');
  }

  const dataset_path = process.argv[2];
  var dataset = JSON.parse(fs.readFileSync(dataset_path, 'utf8'));
  var data = {};

  for (let rec of dataset) {
    let rtype = rec.model.split('.')[1];
    let datum = rec.fields;
    if (rec.pk !== undefined) {
      datum.id = rec.pk;
    }
    // convert datetime
    if (rtype == 'review') {
      datum.creation_time = new Date(datum.creation_time);
    }

    if (!(rtype in data)) {
      data[rtype] = [];
    }
    data[rtype].push(datum);
  }

  await bulk_insert(app, data, 'User');
  await bulk_insert(app, data, 'Person');
  await bulk_insert(app, data, 'Movie');
  await bulk_insert(app, data, 'Review');
  await bulk_insert(app, data, 'Directors');
  await bulk_insert(app, data, 'Cast');

  console.log('Models created.');
});
